### DISCUSSION TITLE
how to extrapolate between two masters?

### DISCUSSION THREAD

**Discussion by aminabedi68** [fonttools/fonttools#3226] [Q&A] [answered]
Hi
(I know the question is more likely about fontmake, but i guess i have more chance to get a clear answer here)
I use fontmake(ufo(from fontforge) + designspace) to make variable fonts in some projects. fontforge has it's own interpolate function, working for both interpolate and extrapolate, but couldn't generate opentype for that at all. fontmake could interpolate(with all opentype features support) but no extrapolate allowed. i tried to force it to use MutatorMath instead of varLib, but that part not working apparently. have you any idea how can i extrapolate between 2 masters(1d) with recalculated GPOS opentype features?

**Comment by behdad** [comment, marked as answer]
We don't support that currently. It's not part of the OpenType model, so if we ever implement it, we have to synthesize the resulting model back into OpenType's.

**Comment by aminabedi68** [comment]
OK. thank you @behdad
I thought that working inside [0:1] (in normalized values) ​​was a secondary limitation and we could hack it somehow(with negative and above 1 values)

**Comment by anthrotype** [comment]
I think Amin was asking about generating instances via extrapolation, not building a VF (in which extrapolation is not contemplated by OT model), am I right?

@LettError's https://github.com/LettError/ufoProcessor (successor of MutatorMath) maybe might help you with this.

The FontTools' VariationModel now has an option extrapolate=True (since https://github.com/fonttools/fonttools/pull/2757), we used that for example in glyphsLib Smart Components support. I think ufoProcessor uses that one too for its extrapolation support.

Regarding Amin's ask about extrapolating "GPOS opentype features", maybe something like our [fontTools.varLib.interpolate_layout](https://github.com/fonttools/fonttools/blob/main/Lib/fontTools/varLib/interpolate_layout.py) could be modified to allow optional extrapolation support via VariationModel extrapolate=True?

**Reply by aminabedi68** [reply]
thanks @anthrotype , i thought it was clear, my bad.
let me explain in more details: with fontmake i can move masters **inside** designspace, by simply defining an instance and(with some preparation) using it as new master. I need extrapolation to moving masters locations to the **outside** of the designspace borders. the main problems (for me) are to get kerning values and anchor marks coordinations.

**Reply by LettError** [reply]
Mutatormath can extrapolate glyphs, kerning, all the geometry. UFOprocessor can, using MutatorMath, generate instance UFOs on extrapolating locations that fontmake will ignore. Let me know if you want an example.

**Reply by aminabedi68** [reply]
Hi @LettError 
sounds great. yes please.

**Reply by LettError** [reply]
This is a script that shows how to generate an instance UFO at an extrapolated location:
https://github.com/LettError/mutatorSans/blob/master/extrapolation_example.py

..using this designspace:
https://github.com/LettError/mutatorSans/blob/master/MutatorSans-weight-only-extrapolating.designspace

Just get the whole mutatorSans repository to get all the source fonts as well. 

The script uses UFOProcessor.UFOOperator to generate the instance UFOs:
https://github.com/LettError/ufoProcessor/blob/master/Lib/ufoProcessor/ufoOperator.py
It has some useful flags: it can use either varlib or mutatorMath, as these extrapolate differently. And it can clip (basically what variable fonts do, past the extremes there are no changes) or extrapolate (the shapes continue to grow).

**Reply by aminabedi68** [reply]
thank you. It's working nice 👌
the Anchor points coordination extrapolated correctly.

![1](https://github.com/fonttools/fonttools/assets/25493297/4b358dba-6cb5-4d0f-8c09-72b7a88d2be1)
